examples/blockchain/network.rs

```diff
@@ -0,0 +1,377 @@
+            .map_err(|_| ())
+    }
+    fn broadcast(&self, data: NetworkData) -> Result<(), Self::Error> {
+        self.msg_to_manager_tx
+            .unbounded_send((data.encode(), None))
+            .map_err(|_| ())
+    }
+    async fn next_event(&mut self) -> Option<NetworkData> {
+        self.msg_from_manager_rx.next().await.map(|msg| {
+            NetworkData::decode(&mut &msg[..]).expect("honest network data should decode")
+        })
+    }
+}
+
+pub(crate) struct NetworkManager {
+    swarm: Swarm<Behaviour>,
+    consensus_rx: mpsc::UnboundedReceiver<(Vec<u8>, Option<NodeIndex>)>,
+    block_rx: mpsc::UnboundedReceiver<Block>,
+}
+
+impl Network {
+    pub(crate) async fn new(
+        node_ix: NodeIndex,
+    ) -> Result<
+        (
+            Self,
+            NetworkManager,
+            UnboundedSender<Block>,
+            UnboundedReceiver<Block>,
+        ),
+        Box<dyn Error>,
+    > {
+        let local_key = identity::Keypair::generate_ed25519();
+        let local_peer_id = PeerId::from(local_key.public());
+        info!(target: "Blockchain-network", "Local peer id: {:?}", local_peer_id);
+
+        let transport = development_transport(local_key).await?;
+
+        let (msg_to_manager_tx, msg_to_manager_rx) = mpsc::unbounded();
+        let (msg_from_manager_tx, msg_from_manager_rx) = mpsc::unbounded();
+        let (block_to_data_io_tx, block_to_data_io_rx) = mpsc::unbounded();
+        let (block_from_data_io_tx, block_from_data_io_rx) = mpsc::unbounded();
+        let mut swarm = {
+            let mut rr_cfg = RequestResponseConfig::default();
+            rr_cfg.set_connection_keep_alive(Duration::from_secs(10));
+            rr_cfg.set_request_timeout(Duration::from_secs(4));
+            let protocol_support = ProtocolSupport::Full;
+            // ProtocolSupport::Outbound
+            let mdns_config = MdnsConfig {
+                ttl: Duration::from_secs(6 * 60),
+                query_interval: Duration::from_millis(100),
+            };
+            let rq_rp = RequestResponse::new(
+                GenericCodec {
+                    max_request_size: 1 << 26,
+                    max_response_size: 1 << 26,
+                },
+                iter::once((ALEPH_PROTOCOL_NAME.as_bytes().to_vec(), protocol_support)),
+                rr_cfg,
+            );
+
+            let mdns = Mdns::new(mdns_config).await?;
+            let behaviour = Behaviour {
+                rq_rp,
+                mdns,
+                peers: vec![],
+                peer_by_index: HashMap::new(),
+                consensus_tx: msg_from_manager_tx,
+                block_tx: block_to_data_io_tx,
+                node_ix,
+            };
+            SwarmBuilder::new(transport, behaviour, local_peer_id)
+                .executor(Box::new(|fut| {
+                    tokio::spawn(fut);
+                }))
+                .build()
+        };
+
+        swarm.listen_on("/ip4/0.0.0.0/tcp/0".parse()?)?;
```

Comment:
Oh, just IPv4? Is v6 not worth the effort, unlikely to be supported by the host or some other reason? This is not a problem, just curiosity.

Reply:
This current version works on localhost only so I didn't spent much time thinking about this line (actually I spent no time on it as I copied it from Mike's dummy-honest.rs :) ). Making it work over WAN will be some fun -- we can consider IPv6 then.